lib/Dialect/Arc/Transforms/SinkInputs.cpp

```diff
@@ -0,0 +1,134 @@
+//===- SinkInputs.cpp------ -----------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "arc-sink-inputs"
+
+using namespace circt;
+using namespace arc;
+using namespace hw;
+
+namespace {
+struct SinkInputsPass : public SinkInputsBase<SinkInputsPass> {
+  void runOnOperation() override;
+  void runOnModule();
+};
+} // namespace
+
+void SinkInputsPass::runOnOperation() {
+  // For each arc, this map tracks which arguments are constant.
+  DenseMap<StringRef, SmallVector<ConstantOp>> arcConstArgs;
+  // All stateOps that use a particular arc.
+  DenseMap<StringRef, SmallVector<StateOp>> arcUses;
+
+  auto module = getOperation();
+  for (auto moduleOp :
+       llvm::make_early_inc_range(module.getOps<HWModuleOp>())) {
+    for (auto op : moduleOp.getBody().getOps<StateOp>()) {
+      // Find all stateOps that use constant operands.
+      auto ops = op.getInputs();
+      SmallVector<ConstantOp> constArgs(ops.size());
```

Comment:
Perhaps you want to allow ConstantLike so that we can sink aggregaet constant ops ( which I believe generated a lot by lut op).

Reply:
Good point! Changed :+1: